### PR TITLE
TypeVisitor: use `std::ops::ControlFlow` instead of `bool`

### PR DESCRIPTION
**Pull request by LeSeulArtichaut**
Implements MCP rust-lang/compiler-team#374.

Blocked on FCP in rust-lang/compiler-team#374.
r? @lcnr cc @jonas-schievink

### CODE REVIEW

File: compiler/rustc_middle/src/macros.rs [hunk 1]
```diff
@@ -62,9 +62,9 @@ macro_rules! CloneTypeFoldableImpls {
                 fn super_visit_with<F: $crate::ty::fold::TypeVisitor<$tcx>>(
                     &self,
                     _: &mut F)
-                    -> bool
+                    -> ::std::ops::ControlFlow<(), ()>
                 {
-                    false
+                    ::std::ops::ControlFlow::CONTINUE
```
Comment (lcnr) [inline compiler/rustc_middle/src/macros.rs:67]: is there a reason for using `::std::ops::ControlFlow` here and `ControlFlow` further down?

File: compiler/rustc_middle/src/mir/type_foldable.rs [hunk 2]
```diff
@@ -88,41 +88,46 @@ impl<'tcx> TypeFoldable<'tcx> for Terminator<'tcx> {
...
+                if let Some((ref loc, _)) = *destination {
+                    loc.visit_with(visitor)?;
                 };
-                dest || func.visit_with(visitor) || args.visit_with(visitor)
+                func.visit_with(visitor)?;
+                args.visit_with(visitor)
             }
             Assert { ref cond, ref msg, .. } => {
-                if cond.visit_with(visitor) {
+                if cond.visit_with(visitor) == ControlFlow::BREAK {
```
Comment (lcnr) [inline compiler/rustc_middle/src/mir/type_foldable.rs:113]: that's weird, why are we only visiting here if `cond.visit_with(...)` returns true?

Is this a bug?

File: compiler/rustc_middle/src/ty/fold.rs [hunk 3]
```diff
@@ -74,7 +76,7 @@ pub trait TypeFoldable<'tcx>: fmt::Debug + Clone {
     }
 
     fn has_type_flags(&self, flags: TypeFlags) -> bool {
-        self.visit_with(&mut HasTypeFlagsVisitor { flags })
+        self.visit_with(&mut HasTypeFlagsVisitor { flags }) == ControlFlow::Break(())
```
Comment (lcnr) [inline compiler/rustc_middle/src/ty/fold.rs:79]: Yeah, it might make sense to introduce a newtype here in the future, so this can be `visit_with(...).break_value() == Some(FoundFlags)`.

Not part of this PR though

File: compiler/rustc_middle/src/ty/fold.rs [hunk 4]
```diff
@@ -300,7 +305,8 @@ impl<'tcx> TyCtxt<'tcx> {
         value: &impl TypeFoldable<'tcx>,
         callback: impl FnMut(ty::Region<'tcx>) -> bool,
     ) -> bool {
-        return value.visit_with(&mut RegionVisitor { outer_index: ty::INNERMOST, callback });
+        return value.visit_with(&mut RegionVisitor { outer_index: ty::INNERMOST, callback })
+            == ControlFlow::BREAK;
```
Comment (lcnr) [inline compiler/rustc_middle/src/ty/fold.rs:309]: pre existing, but can you move that expression to the end of that function and remove the `return`?

This was a bit confusing for me

File: compiler/rustc_middle/src/ty/fold.rs [hunk 5]
```diff
@@ -328,28 +334,32 @@ impl<'tcx> TyCtxt<'tcx> {
...
                 self.outer_index.shift_in(1);
                 let result = t.as_ref().skip_binder().visit_with(self);
                 self.outer_index.shift_out(1);
                 result
             }
 
-            fn visit_region(&mut self, r: ty::Region<'tcx>) -> bool {
-                match *r {
+            fn visit_region(&mut self, r: ty::Region<'tcx>) -> ControlFlow<(), ()> {
+                let should_break = match *r {
```
Comment (lcnr) [inline compiler/rustc_middle/src/ty/fold.rs:348]: I think it's better to also change `self.callback` to return `ControlFlow` here.

File: compiler/rustc_middle/src/ty/fold.rs [hunk 6]
```diff
@@ -834,45 +844,57 @@ struct HasEscapingVarsVisitor {
...
-    fn visit_ty(&mut self, t: Ty<'tcx>) -> bool {
+    fn visit_ty(&mut self, t: Ty<'tcx>) -> ControlFlow<(), ()> {
         // If the outer-exclusive-binder is *strictly greater* than
         // `outer_index`, that means that `t` contains some content
         // bound at `outer_index` or above (because
         // `outer_exclusive_binder` is always 1 higher than the
         // content in `t`). Therefore, `t` has some escaping vars.
-        t.outer_exclusive_binder > self.outer_index
+        if t.outer_exclusive_binder > self.outer_index {
+            ControlFlow::BREAK
```
Comment (lcnr) [inline compiler/rustc_middle/src/ty/fold.rs:861]: yeah, looking at this I really like the idea of using a newtype for `ControlFlow::Break(...)` here.

Still not relevant to this PR though

File: compiler/rustc_middle/src/ty/structural_impls.rs [hunk 7]
```diff
@@ -820,8 +824,11 @@ impl<'tcx, T: TypeFoldable<'tcx>> TypeFoldable<'tcx> for Vec<T> {
         self.iter().map(|t| t.fold_with(folder)).collect()
     }
 
-    fn super_visit_with<V: TypeVisitor<'tcx>>(&self, visitor: &mut V) -> bool {
-        self.iter().any(|t| t.visit_with(visitor))
+    fn super_visit_with<V: TypeVisitor<'tcx>>(&self, visitor: &mut V) -> ControlFlow<(), ()> {
+        for t in self {
```
Comment (lcnr) [inline compiler/rustc_middle/src/ty/structural_impls.rs:828]: can you use `self.iter().try_for_each` here?

File: compiler/rustc_mir/src/interpret/util.rs [hunk 8]
```diff
@@ -50,10 +51,10 @@ where
                         match (is_used, subst.needs_subst()) {
                             // Just in case there are closures or generators within this subst,
                             // recurse.
-                            (true, true) if subst.super_visit_with(self) => {
+                            (true, true) if subst.super_visit_with(self) == ControlFlow::BREAK => {
                                 // Only return when we find a parameter so the remaining substs
                                 // are not skipped.
-                                return true;
+                                return ControlFlow::BREAK;
```
Comment (lcnr) [inline compiler/rustc_mir/src/interpret/util.rs:57]: Isn't this simply
```rust
(true, true) => subst.super_visit_with(self),
```

File: compiler/rustc_privacy/src/lib.rs [hunk 9]
```diff
@@ -77,54 +79,54 @@ impl<'tcx, V> DefIdVisitorSkeleton<'_, 'tcx, V>
 where
     V: DefIdVisitor<'tcx> + ?Sized,
 {
-    fn visit_trait(&mut self, trait_ref: TraitRef<'tcx>) -> bool {
+    fn visit_trait(&mut self, trait_ref: TraitRef<'tcx>) -> ControlFlow<(), ()> {
         let TraitRef { def_id, substs } = trait_ref;
-        self.def_id_visitor.visit_def_id(def_id, "trait", &trait_ref.print_only_trait_path())
-            || (!self.def_id_visitor.shallow() && substs.visit_with(self))
+        if self.def_id_visitor.visit_def_id(def_id, "trait", &trait_ref.print_only_trait_path()) {
```
Comment (lcnr) [inline compiler/rustc_privacy/src/lib.rs:84]: should `visit_def_id` potentionally also return `ControlFlow`? :thinking:

File: compiler/rustc_privacy/src/lib.rs [hunk 10]
```diff
@@ -77,54 +79,54 @@ impl<'tcx, V> DefIdVisitorSkeleton<'_, 'tcx, V>
...
     }
 
-    fn visit_predicates(&mut self, predicates: ty::GenericPredicates<'tcx>) -> bool {
+    fn visit_predicates(&mut self, predicates: ty::GenericPredicates<'tcx>) -> ControlFlow<(), ()> {
         let ty::GenericPredicates { parent: _, predicates } = predicates;
         for &(predicate, _span) in predicates {
-            if self.visit_predicate(predicate) {
-                return true;
-            }
+            self.visit_predicate(predicate)?;
```
Comment (lcnr) [inline compiler/rustc_privacy/src/lib.rs:119]: maybe `try_for_each`

File: compiler/rustc_privacy/src/lib.rs [hunk 11]
```diff
@@ -229,7 +225,11 @@ where
             }
         }
 
-        !self.def_id_visitor.shallow() && ty.super_visit_with(self)
+        if !self.def_id_visitor.shallow() {
```
Comment (lcnr) [inline compiler/rustc_privacy/src/lib.rs:228]: please swap the if branches here and use `if self.def_id_visitor.shallow()`

File: compiler/rustc_privacy/src/lib.rs [hunk 12]
```diff
@@ -1193,12 +1193,14 @@ impl<'tcx> TypePrivacyVisitor<'tcx> {
     fn check_expr_pat_type(&mut self, id: hir::HirId, span: Span) -> bool {
         self.span = span;
         let typeck_results = self.typeck_results();
-        if self.visit(typeck_results.node_type(id)) || self.visit(typeck_results.node_substs(id)) {
+        if self.visit(typeck_results.node_type(id)) == ControlFlow::BREAK
```
Comment (lcnr) [inline compiler/rustc_privacy/src/lib.rs:1196]: maybe use a `try` block here instead of having to manually check for `ControlFlow::BREAK` here

File: compiler/rustc_trait_selection/src/traits/structural_match.rs [hunk 13]
```diff
@@ -258,16 +248,16 @@ impl<'a, 'tcx> TypeVisitor<'tcx> for Search<'a, 'tcx> {
             let ty = self.tcx().normalize_erasing_regions(ty::ParamEnv::empty(), field_ty);
             debug!("structural-match ADT: field_ty={:?}, ty={:?}", field_ty, ty);
 
-            if ty.visit_with(self) {
+            if ty.visit_with(self) == ControlFlow::BREAK {
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/traits/structural_match.rs:251]: yeah, we probably want to store `self.found` in `ControlFlow::Break` here.

(also not part of this PR)

File: compiler/rustc_typeck/src/check/wfcheck.rs [hunk 14]
```diff
@@ -798,26 +800,26 @@ fn check_where_clauses<'tcx, 'fcx>(
                 params: FxHashSet<u32>,
             }
             impl<'tcx> ty::fold::TypeVisitor<'tcx> for CountParams {
```
Comment (lcnr) [inline compiler/rustc_typeck/src/check/wfcheck.rs:802]: That folder seems weird to me, why are we storing the params in a hashmap if we only care about if there is more than one of them?

Using an option is probably more appropriate here but whatever :shrug: Isn't really relevant for this PR

File: src/tools/clippy/clippy_lints/src/redundant_clone.rs [hunk 15]
```diff
@@ -558,8 +559,8 @@ impl<'a, 'tcx> mir::visit::Visitor<'tcx> for PossibleBorrowerVisitor<'a, 'tcx> {
 struct ContainsRegion;
 
 impl TypeVisitor<'_> for ContainsRegion {
-    fn visit_region(&mut self, _: ty::Region<'_>) -> bool {
-        true
+    fn visit_region(&mut self, _: ty::Region<'_>) -> ControlFlow<(), ()> {
+        ControlFlow::BREAK
```
Comment (lcnr) [inline src/tools/clippy/clippy_lints/src/redundant_clone.rs:563]: Interesting that we aren't ignoring `ReStatic` here.

**Review by lcnr — APPROVED** [review]
Thanks.

Some nitpicks and suggestion but this is already looking a lot clearer to me :heart:

I do have to admit that my brain ended up skipping over some `bool`  -> `ControlFlow` conversions so hopefully I didn't accidentially miss anything :sweat_smile:

File: compiler/rustc_middle/src/mir/type_foldable.rs (hunk 2 above)
Comment (oli-obk) [reply compiler/rustc_middle/src/mir/type_foldable.rs:113]: yea this looks wrong. I don't know why it would be this way around. I mean this is essentially `cond.visit_with(visitor) && {...}` which is not something that is ever used anywhere

File: compiler/rustc_middle/src/ty/fold.rs (hunk 5 above)
Comment (LeSeulArtichaut) [reply compiler/rustc_middle/src/ty/fold.rs:348]: This is for `any_free_region_meets`, I think it makes more sense to require `bool` so it can be used like `Iterator::any`.

Comment (lcnr) [reply compiler/rustc_middle/src/ty/fold.rs:348]: hmm, makes sense :thinking: 

Can you instead move the `if should_break` into the match arm and directly return `ControlFlow::CONTINUE` in the `ReLateBound` case?

File: compiler/rustc_privacy/src/lib.rs (hunk 9 above)
Comment (LeSeulArtichaut) [reply compiler/rustc_privacy/src/lib.rs:84]: Yes, I think it should

File: compiler/rustc_privacy/src/lib.rs (hunk 12 above)
Comment (LeSeulArtichaut) [reply compiler/rustc_privacy/src/lib.rs:1193]: You mean something like?

```rust
if try {
    self.visit(typeck_results.node_type(id))?;
    self.visit(typeck_results.node_substs(id))
} == ControlFlow::BREAK {
    // ...
}
```

File: compiler/rustc_middle/src/mir/type_foldable.rs [hunk 16]
```diff
@@ -164,8 +170,11 @@ impl<'tcx> TypeFoldable<'tcx> for &'tcx ty::List<PlaceElem<'tcx>> {
         folder.tcx().intern_place_elems(&v)
     }
 
-    fn super_visit_with<V: TypeVisitor<'tcx>>(&self, visitor: &mut V) -> bool {
-        self.iter().any(|t| t.visit_with(visitor))
+    fn super_visit_with<V: TypeVisitor<'tcx>>(&self, visitor: &mut V) -> ControlFlow<(), ()> {
+        for t in self.iter() {
+            t.visit_with(visitor)?;
+        }
+        ControlFlow::CONTINUE
```
Comment (oli-obk) [inline compiler/rustc_middle/src/mir/type_foldable.rs:177]:
```suggestion
        self.iter().map(|t| t.visit_with(visitor)).collect()
```

needs a `FromIterator` impl for `ControlFlow`, it's not automatic for `Try`, right?

File: compiler/rustc_privacy/src/lib.rs (hunk 12 above)
Comment (lcnr) [reply compiler/rustc_privacy/src/lib.rs:1187]: yeah, though probably for the whole body of `check_expr_pat_type` as there are quite a lot of these `if`s

File: compiler/rustc_middle/src/mir/type_foldable.rs (hunk 16 above)
Comment (lcnr) [reply compiler/rustc_middle/src/mir/type_foldable.rs:177]: my idea would be to instead use `self.iter().try_for_each(|v| t.visit_with(visitor))`

Comment (lcnr) [reply compiler/rustc_middle/src/mir/type_foldable.rs:177]: which imo is a bit clearer than implementing `FromIterator` here

File: compiler/rustc_middle/src/mir/type_foldable.rs [hunk 17]
```diff
@@ -214,32 +223,41 @@ impl<'tcx> TypeFoldable<'tcx> for Rvalue<'tcx> {
...
+                    AggregateKind::Tuple => ControlFlow::CONTINUE,
                     AggregateKind::Adt(_, _, substs, user_ty, _) => {
-                        substs.visit_with(visitor) || user_ty.visit_with(visitor)
+                        substs.visit_with(visitor)?;
+                        user_ty.visit_with(visitor)
                     }
                     AggregateKind::Closure(_, substs) => substs.visit_with(visitor),
                     AggregateKind::Generator(_, substs, _) => substs.visit_with(visitor),
-                }) || fields.visit_with(visitor)
+                })?;
```
Comment (oli-obk) [inline compiler/rustc_middle/src/mir/type_foldable.rs:259]: I think this match should be adjusted to use `?` inside instead of being an expression that `?` is applied to

File: compiler/rustc_middle/src/mir/type_foldable.rs (hunk 16 above)
Comment (oli-obk) [reply compiler/rustc_middle/src/mir/type_foldable.rs:177]: yea, I've seen this is actually done in other places in the PR. I did not know about `try_for_each`, and it does indeed seem more applicable here.

File: compiler/rustc_middle/src/ty/subst.rs [hunk 18]
```diff
@@ -391,8 +392,11 @@ impl<'tcx> TypeFoldable<'tcx> for SubstsRef<'tcx> {
         }
     }
 
-    fn super_visit_with<V: TypeVisitor<'tcx>>(&self, visitor: &mut V) -> bool {
-        self.iter().any(|t| t.visit_with(visitor))
+    fn super_visit_with<V: TypeVisitor<'tcx>>(&self, visitor: &mut V) -> ControlFlow<(), ()> {
+        for t in *self {
+            t.visit_with(visitor)?;
+        }
+        ControlFlow::CONTINUE
```
Comment (oli-obk) [inline compiler/rustc_middle/src/ty/subst.rs:399]: another `try_for_each` candidate

File: compiler/rustc_mir/src/monomorphize/polymorphize.rs [hunk 19]
```diff
@@ -317,28 +318,38 @@ struct HasUsedGenericParams<'a> {
...
         debug!("visit_const: c={:?}", c);
         if !c.has_param_types_or_consts() {
-            return false;
+            return ControlFlow::CONTINUE;
         }
 
         match c.val {
             ty::ConstKind::Param(param) => {
-                !self.unused_parameters.contains(param.index).unwrap_or(false)
+                if !self.unused_parameters.contains(param.index).unwrap_or(false) {
```
Comment (oli-obk) [inline compiler/rustc_mir/src/monomorphize/polymorphize.rs:329]: swap the branches and invert the condition here

File: compiler/rustc_mir/src/monomorphize/polymorphize.rs [hunk 20]
```diff
@@ -317,28 +318,38 @@ struct HasUsedGenericParams<'a> {
...
         debug!("visit_ty: ty={:?}", ty);
         if !ty.has_param_types_or_consts() {
-            return false;
+            return ControlFlow::CONTINUE;
         }
 
         match ty.kind() {
-            ty::Param(param) => !self.unused_parameters.contains(param.index).unwrap_or(false),
+            ty::Param(param) => {
+                if !self.unused_parameters.contains(param.index).unwrap_or(false) {
```
Comment (oli-obk) [inline compiler/rustc_mir/src/monomorphize/polymorphize.rs:347]: swap and invert

File: compiler/rustc_privacy/src/lib.rs [hunk 21]
```diff
@@ -77,54 +79,49 @@ impl<'tcx, V> DefIdVisitorSkeleton<'_, 'tcx, V>
 where
     V: DefIdVisitor<'tcx> + ?Sized,
 {
-    fn visit_trait(&mut self, trait_ref: TraitRef<'tcx>) -> bool {
+    fn visit_trait(&mut self, trait_ref: TraitRef<'tcx>) -> ControlFlow<(), ()> {
         let TraitRef { def_id, substs } = trait_ref;
-        self.def_id_visitor.visit_def_id(def_id, "trait", &trait_ref.print_only_trait_path())
-            || (!self.def_id_visitor.shallow() && substs.visit_with(self))
+        self.def_id_visitor.visit_def_id(def_id, "trait", &trait_ref.print_only_trait_path())?;
+        if !self.def_id_visitor.shallow() { substs.visit_with(self) } else { ControlFlow::CONTINUE }
```
Comment (oli-obk) [inline compiler/rustc_privacy/src/lib.rs:85]: swap and invert

File: compiler/rustc_privacy/src/lib.rs [hunk 22]
```diff
@@ -1407,8 +1401,12 @@ impl DefIdVisitor<'tcx> for TypePrivacyVisitor<'tcx> {
     fn tcx(&self) -> TyCtxt<'tcx> {
         self.tcx
     }
-    fn visit_def_id(&mut self, def_id: DefId, kind: &str, descr: &dyn fmt::Display) -> bool {
-        self.check_def_id(def_id, kind, descr)
+    fn visit_def_id(&mut self, def_id: DefId, kind: &str, descr: &dyn fmt::Display) -> ControlFlow<(), ()> {
+        if self.check_def_id(def_id, kind, descr) {
```
Comment (oli-obk) [inline compiler/rustc_privacy/src/lib.rs:1405]: these are very common. Maybe we should have a `ControlFlow::break_if(some_bool)` and `ControlFlow::continue_if(some_bool)` for convenience? Or will that become infeasible with types other than `()`?

File: compiler/rustc_trait_selection/src/opaque_types.rs [hunk 23]
```diff
@@ -691,26 +693,26 @@ impl<'tcx, OP> TypeVisitor<'tcx> for ConstrainOpaqueTypeRegionVisitor<OP>
 where
     OP: FnMut(ty::Region<'tcx>),
 {
-    fn visit_binder<T: TypeFoldable<'tcx>>(&mut self, t: &ty::Binder<T>) -> bool {
+    fn visit_binder<T: TypeFoldable<'tcx>>(&mut self, t: &ty::Binder<T>) -> ControlFlow<(), ()> {
         t.as_ref().skip_binder().visit_with(self);
```
Comment (oli-obk) [inline compiler/rustc_trait_selection/src/opaque_types.rs:697]: `ControlFlow` should probably be `#[must_use]` so we'd be forced to use `let _ = ` here, but that's for another PR

File: compiler/rustc_typeck/src/check/check.rs [hunk 24]
```diff
@@ -417,30 +419,34 @@ pub(super) fn check_opaque_for_inheriting_lifetimes(
     };
 
     impl<'tcx> ty::fold::TypeVisitor<'tcx> for ProhibitOpaqueVisitor<'tcx> {
-        fn visit_ty(&mut self, t: Ty<'tcx>) -> bool {
+        fn visit_ty(&mut self, t: Ty<'tcx>) -> ControlFlow<(), ()> {
             debug!("check_opaque_for_inheriting_lifetimes: (visit_ty) t={:?}", t);
-            if t != self.opaque_identity_ty && t.super_visit_with(self) {
+            if t != self.opaque_identity_ty && t.super_visit_with(self) == ControlFlow::BREAK {
```
Comment (oli-obk) [inline compiler/rustc_typeck/src/check/check.rs:424]: This alos looks like a candidate for returning a value via `Break`

**Comment by oli-obk** [comment]
This PR is awesome (and exposes so many issues of different sublety

File: compiler/rustc_trait_selection/src/opaque_types.rs (hunk 23 above)
Comment (lcnr) [reply compiler/rustc_trait_selection/src/opaque_types.rs:697]: Yeah, making it `#[must_use]` seems like a good idea to me

File: compiler/rustc_privacy/src/lib.rs (hunk 22 above)
Comment (lcnr) [reply compiler/rustc_privacy/src/lib.rs:1405]: yeah, I think adding these methods make sense. I am not yet sure how to be generalize them to deal with types other than `ControlFlow<(), ()>`

Comment (scottmcm) [reply compiler/rustc_privacy/src/lib.rs:1405]: There has been [almost no thought](https://github.com/rust-lang/rust/issues/75744#issuecomment-679250842) put into what methods the `ControlFlow` type should have, so if you come up with reasonable ones definitely add them.  It used to be an internal implementation detail, so doesn't even have the usual combinators.

Comment (LeSeulArtichaut) [reply compiler/rustc_privacy/src/lib.rs:1425]: @scottmcm I'd say definitely `is_break` and `is_continue` (name to be bikeshed of course). I'll file a PR

File: compiler/rustc_trait_selection/src/opaque_types.rs (hunk 23 above)
Comment (LeSeulArtichaut) [reply compiler/rustc_trait_selection/src/opaque_types.rs:697]: Opened #78202

Comment (scottmcm) [reply compiler/rustc_trait_selection/src/opaque_types.rs:697]: Another thing you could consider to avoid the conversation on the type would be to just mark the methods on the trait as `#[must_use]`.  It [looks](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=da1fcb2728ab2129196d98af35b367d9) like that works even for calls on concrete things that don't copy the attribute to their implementation.  Probably still not this PR, though.

File: compiler/rustc_middle/src/ty/structural_impls.rs [hunk 25]
```diff
@@ -1075,11 +1090,13 @@ impl<'tcx> TypeFoldable<'tcx> for &'tcx ty::Const<'tcx> {
         folder.fold_const(*self)
     }
 
-    fn super_visit_with<V: TypeVisitor<'tcx>>(&self, visitor: &mut V) -> bool {
-        self.ty.visit_with(visitor) || self.val.visit_with(visitor)
+    fn super_visit_with<V: TypeVisitor<'tcx>>(&self, visitor: &mut V) -> ControlFlow<()> {
+        self.ty.visit_with(visitor)?;
+        self.val.visit_with(visitor)?;
+        ControlFlow::CONTINUE
```
Comment (lcnr) [inline compiler/rustc_middle/src/ty/structural_impls.rs:1096]:
```suggestion
        self.val.visit_with(visitor)?
```

**Comment by LeSeulArtichaut** [comment]
@oli-obk @lcnr The MCP finished its 10 days FCP

**Comment by oli-obk** [comment]
@bory r=lcnr,oli-obk

**Comment by LeSeulArtichaut** [comment]
@oli-obk you made a small typo 👆😁
